src/Addon/Recurring.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace Give\Addon;
+
+use Give_Subscription;
+use Give_Subscriptions_DB;
+use function Give\Helpers\Form\Template\Utils\Frontend\getPaymentId;
+
+/**
+ * Class Addon
+ *
+ * @package Give\Addon
+ */
+class Recurring implements Addonable {
+
+	/**
+	 * @inheritDoc
+	 */
+	public static function isActive() {
+		return defined( 'GIVE_RECURRING_VERSION' );
+	}
+
+	/**
+	 * Get subscription from donation id.
+	 *
+	 * @since 2.7.0
+	 *
+	 * @param int|null $donationID
+	 *
+	 * @return Give_Subscription|null
+	 */
+	public static function getSubscriptionFromInitialDonationId( $donationID ) {
```

Comment:
I don't think this really belongs here. Let's think about the purpose of this class for a moment; what is it's Single Responsibility? It seems like the purpose of this class is to represent an addon, that is to determine if another plugin exists which is extending GiveWP. In this case the `isActive()` method makes perfect sense. This represents the _business_ layer of the application as it has to do with business logic. This method, `getSubscriptionFromInitialDonationId` is a _persistence_ layer method, as you're interacting with the database to retrieve data.

So this class is crossing multiple layers of logic and doing two every different things. Is it for representing a GiveWP addon? Or for retrieving data from the database? I understand the thinking "well, its the recurring addon, so this does recurring stuff", but that's too broad. Recurring, in GiveWP, would be considered a "domain" of logic — an umbrella for different kinds of logic — but this would never be represented as a single class, but a group of classes.

This would go in a `RecurringRepository` class, whose Single Responsibility is to port data between the persistence and business layer. It's nice that the business layer can say "hey, I want the subscription for this donation id, and I don't care how you get it", but that's the job of a repository.

Make sense?

cc: @henryholtgeerts

Reply:
This code has been removed